File: src/client/java/com/rafalohaki/module/modules/KillauraModule.java
package com.rafalohaki.module.modules;

import com.rafalohaki.event.ClientTickEvent;
import com.rafalohaki.event.EventBus;
import com.rafalohaki.module.Category;
import com.rafalohaki.module.Module;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.ClientPlayNetworkHandler;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.network.packet.c2s.play.HandSwingC2SPacket;
import net.minecraft.network.packet.c2s.play.PlayerInteractEntityC2SPacket;
import net.minecraft.network.packet.c2s.play.PlayerMoveC2SPacket;
import net.minecraft.util.Hand;
import net.minecraft.util.math.Box;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.RaycastContext;

import java.util.List;
import java.util.Random;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Advanced Killaura with proper flick-based attack pattern
 * - Single rotation packet per attack (flick technique)
 * - Random intervals between decisions
 * - Human-like aim with noise and jitter
 * - No state machine - simple conditional logic
 */
public class KillauraModule extends Module {
    private float range = 3.2f;
    private float cps = 10f;
    private boolean playersOnly = false;
    private boolean throughWalls = false;
    
    // Stan wewnętrzny, proste flagi, nie skomplikowana maszyna
    private final AtomicReference<LivingEntity> currentTarget = new AtomicReference<>(null);
    private long lastAttackTime = 0L;
    private long nextAttackDelay = 1000L; // Zostanie obliczone w onEnable
    private int tickCounter = 0;
    private int nextActionTick = 2; // Losowy interwał, np. 2-4 ticki

    private final Random random = new Random();
    private final ClientTickEvent.Handler tickHandler = event -> onTick(event);
    
    public KillauraModule() {
        super("Killaura", "Advanced combat with proper flick technique", Category.COMBAT);
    }
    
    @Override
    protected void onEnable() {
        lastAttackTime = 0L;
        nextAttackDelay = calculateRandomDelay();
        tickCounter = 0;
        nextActionTick = random.nextInt(2, 5);
    }
    
    @Override
    protected void onDisable() {
        currentTarget.set(null);
    }
    
    @Override
    protected void registerEvents() {
        EventBus.getInstance().register(tickHandler);
    }
    
    @Override
    protected void unregisterEvents() {
        EventBus.getInstance().unregister(tickHandler);
    }
    
    private void onTick(ClientTickEvent event) {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc.player == null || mc.world == null) return;

        // Losowe ticki - psujemy automatyzm
        tickCounter++;
        if (tickCounter < nextActionTick) {
            return; // Nic nie rób w tym ticku
        }
        tickCounter = 0;
        nextActionTick = random.nextInt(2, 5); // Ustaw następny losowy interwał

        // Scan for targets in main thread instead of background
        if (tickCounter == 0) { // Only scan occasionally
            scanForTargets();
        }
        
        LivingEntity target = currentTarget.get();
        if (target == null || !isValidTarget(target)) {
            currentTarget.set(null);
            return;
        }

        // Prosty warunek ataku, bez dodatkowych stanów
        float cooldown = mc.player.getAttackCooldownProgress(0.5f);
        long timeSinceLastAttack = System.currentTimeMillis() - lastAttackTime;

        if (cooldown >= 0.9f && timeSinceLastAttack >= nextAttackDelay) {
            // Walidacja zasięgu PRZED atakiem
            if (mc.player.distanceTo(target) <= range) {
                // Wykonaj całą akcję w jednej funkcji
                performFlickAttack(target); 
                
                lastAttackTime = System.currentTimeMillis();
                nextAttackDelay = calculateRandomDelay(); // Ustaw nowy losowy czas do ataku
            }
        }
    }
    
    // Prosta funkcja zamiast 1000/cps
    private long calculateRandomDelay() {
        float avgDelay = 1000f / cps;
        float variance = avgDelay * 0.25f; // +/- 25%
        float randomOffset = (random.nextFloat() * 2f - 1f) * variance;
        return (long) (avgDelay + randomOffset);
    }
    
    private void performFlickAttack(Entity target) {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc.player == null || mc.getNetworkHandler() == null) return;

        // --- Krok 1: Prognozowanie ---
        // Pobieramy ping w tickach, uproszczony model
        int pingInTicks = (mc.getNetworkHandler().getPlayerListEntry(target.getUuid()) != null) 
            ? mc.getNetworkHandler().getPlayerListEntry(target.getUuid()).getLatency() / 50 
            : 2; // Zakładamy 2 ticki, jeśli nie mamy pinga
        float predictionTimeTicks = pingInTicks + 1; // Dodajemy 1 tick bufora na serwer

        // Oblicz przewidywaną pozycję
        Vec3d predictedPos = new Vec3d(
            target.getX() + target.getVelocity().x * predictionTimeTicks,
            target.getY() + target.getVelocity().y * predictionTimeTicks,
            target.getZ() + target.getVelocity().z * predictionTimeTicks
        );

        // --- Krok 2: Obliczanie rotacji do przewidzianej pozycji ---
        Vec3d eyePos = mc.player.getEyePos();
        // Celuj w losowy punkt na hitboxie celu
        double heightOffset = target.getStandingEyeHeight() * (0.3 + random.nextDouble() * 0.4);
        Vec3d aimPoint = new Vec3d(predictedPos.x, predictedPos.y + heightOffset, predictedPos.z);
        
        Vec3d diff = aimPoint.subtract(eyePos);
        double distanceXZ = Math.sqrt(diff.x * diff.x + diff.z * diff.z);
        
        float idealYaw = (float) Math.toDegrees(Math.atan2(diff.z, diff.x)) - 90f;
        float idealPitch = (float) -Math.toDegrees(Math.atan2(diff.y, distanceXZ));
        
        // --- Krok 3: Dodawanie ludzkiego szumu ---
        // Proste jittery
        float jitterYaw = (random.nextFloat() - 0.5f) * 1.5f;
        float jitterPitch = (random.nextFloat() - 0.5f) * 1.0f;
        
        float finalYaw = normalizeYaw(idealYaw + jitterYaw);
        float finalPitch = Math.max(-90f, Math.min(90f, idealPitch + jitterPitch));

        // --- Krok 4: Wysyłanie pakietów (FLICK!) ---
        // A. Wysyłamy PAKIET Z OBROTEM pozycję i rotację. To jest "flick".
        mc.getNetworkHandler().sendPacket(new PlayerMoveC2SPacket.Full(
            mc.player.getX(), mc.player.getY(), mc.player.getZ(), 
            finalYaw, finalPitch, mc.player.isOnGround()
        ));

        // B. NATYCHMIAST wysyłamy pakiet ATAKU.
        mc.getNetworkHandler().sendPacket(new PlayerInteractEntityC2SPacket(target, mc.player.isSneaking()));

        // C. Opcjonalnie, wymach ręką (losowy)
        if (random.nextFloat() > 0.15f) {
            mc.getNetworkHandler().sendPacket(new HandSwingC2SPacket(Hand.MAIN_HAND));
        }
        
        mc.player.resetLastAttackedTicks();
    }
    
    private void scanForTargets() {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc.player == null || mc.world == null) return;
        
        Vec3d playerPos = mc.player.getPos();
        Box box = new Box(playerPos, playerPos).expand(range * 2, range * 2, range * 2);
        
        List<Entity> entities = mc.world.getOtherEntities(mc.player, box, entity -> 
            entity instanceof LivingEntity && 
            entity.isAlive() && 
            mc.player.distanceTo(entity) <= range &&
            (!playersOnly || entity instanceof PlayerEntity)
        );
        
        LivingEntity closest = null;
        double minDistance = Double.MAX_VALUE;
        
        for (Entity entity : entities) {
            if (entity instanceof LivingEntity && isValidTarget((LivingEntity) entity)) {
                double distance = mc.player.distanceTo(entity);
                if (distance < minDistance) {
                    minDistance = distance;
                    closest = (LivingEntity) entity;
                }
            }
        }
        
        currentTarget.set(closest);
    }
    
    private boolean isValidTarget(LivingEntity target) {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc.player == null) return false;
        
        if (!target.isAlive() || target.getHealth() <= 0f) return false;
        if (mc.player.distanceTo(target) > range) return false;
        if (!throughWalls && !hasLineOfSight(target)) return false;
        
        return true;
    }
    
    private boolean hasLineOfSight(Entity target) {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc.player == null || mc.world == null) return false;
        
        Vec3d eyePos = mc.player.getEyePos();
        Vec3d targetCenter = new Vec3d(
            target.getX(),
            target.getY() + target.getStandingEyeHeight() * 0.5,
            target.getZ()
        );
        Vec3d direction = targetCenter.subtract(eyePos).normalize();
        double distance = eyePos.distanceTo(targetCenter);
        
        RaycastContext context = new RaycastContext(
            eyePos,
            eyePos.add(direction.multiply(distance, distance, distance)),
            RaycastContext.ShapeType.OUTLINE,
            RaycastContext.FluidHandling.NONE,
            mc.player
        );
        
        var result = mc.world.raycast(context);
        return result == null ||
               result.getType() == net.minecraft.util.hit.HitResult.Type.MISS ||
               result.getPos().distanceTo(eyePos) >= distance - 0.1;
    }
    
    private float normalizeYaw(float yaw) {
        float normalized = yaw % 360f;
        if (normalized > 180f) normalized -= 360f;
        if (normalized < -180f) normalized += 360f;
        return normalized;
    }
    
    // Settings getters and setters
    public float getRange() { return range; }
    public float getCps() { return cps; }
    public boolean isPlayersOnly() { return playersOnly; }
    public boolean isThroughWalls() { return throughWalls; }
    
    public void setRange(float newRange) {
        range = Math.max(2.8f, Math.min(3.5f, newRange));
    }
    
    public void setCps(float newCps) {
        cps = Math.max(8f, Math.min(12f, newCps));
    }
    
    public void setPlayersOnly(boolean value) {
        playersOnly = value;
    }
    
    public void setThroughWalls(boolean value) {
        throughWalls = value;
    }
}

File: src/client/kotlin/com/rafalohaki/module/modules/KillauraModule.kt
package com.rafalohaki.module.modules

import com.rafalohaki.event.ClientTickEvent
import com.rafalohaki.event.EventBus
import com.rafalohaki.module.Category
import com.rafalohaki.module.Module
import kotlinx.coroutines.*
import net.minecraft.client.MinecraftClient
import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.network.packet.c2s.play.HandSwingC2SPacket
import net.minecraft.network.packet.c2s.play.PlayerInteractEntityC2SPacket
import net.minecraft.network.packet.c2s.play.PlayerMoveC2SPacket
import net.minecraft.util.Hand
import net.minecraft.util.math.Box
import net.minecraft.util.math.Vec3d
import net.minecraft.world.RaycastContext
import java.util.concurrent.atomic.AtomicReference
import kotlin.math.*
import kotlin.random.Random

class KillauraModule : Module(
    name = "Killaura",
    description = "Humanized combat bot with Grim AC bypass",
    category = Category.COMBAT
) {
    var range = 4.2f
        private set
    var cps = 12f
        private set
    var rotationSpeed = 15f
        private set
    var playersOnly = false
        private set
    var throughWalls = false
        private set
    
    private val currentTarget = AtomicReference<LivingEntity?>(null)
    private val lastAttackTime = AtomicReference(0L)
    private var serverYaw = 0f
    private var serverPitch = 0f
    private var isRotating = false
    
    private enum class State { IDLE, SCANNING, ROTATING, READY, ATTACKING }
    private var currentState = State.IDLE
    private var scanJob: Job? = null
    private val tickHandler: (ClientTickEvent) -> Unit = { onTick() }
    
    override fun onEnable() {
        val mc = MinecraftClient.getInstance()
        serverYaw = mc.player?.yaw ?: 0f
        serverPitch = mc.player?.pitch ?: 0f
        currentState = State.IDLE
        
        scanJob = CoroutineScope(Dispatchers.Default).launch {
            while (isActive) {
                scanForTargets()
                delay(50)
            }
        }
    }
    
    override fun onDisable() {
        scanJob?.cancel()
        currentTarget.set(null)
        currentState = State.IDLE
        isRotating = false
    }
    
    override fun registerEvents() {
        EventBus.register(tickHandler)
    }
    
    override fun unregisterEvents() {
        EventBus.unregister(tickHandler)
    }
    
    private fun onTick() {
        val mc = MinecraftClient.getInstance()
        val player = mc.player ?: return
        val target = currentTarget.get()
        
        when (currentState) {
            State.IDLE -> {
                if (target != null) currentState = State.SCANNING
            }
            State.SCANNING -> {
                if (target == null || !isValidTarget(target)) {
                    currentState = State.IDLE
                    return
                }
                currentState = State.ROTATING
            }
            State.ROTATING -> {
                if (target == null || !isValidTarget(target)) {
                    currentState = State.IDLE
                    return
                }
                
                updateRotations(target)
                sendRotations()
                
                val (idealYaw, idealPitch) = calculateIdealRotations(target)
                val yawDelta = abs(normalizeYaw(idealYaw - serverYaw))
                val pitchDelta = abs(idealPitch - serverPitch)
                
                if (yawDelta < 5f && pitchDelta < 5f) {
                    currentState = State.READY
                }
            }
            State.READY -> {
                if (target == null || !isValidTarget(target)) {
                    currentState = State.IDLE
                    return
                }
                
                updateRotations(target)
                sendRotations()
                
                val cooldown = player.getAttackCooldownProgress(0.5f)
                val timeSinceLastAttack = System.currentTimeMillis() - lastAttackTime.get()
                val minDelay = (1000f / cps).toLong()
                
                if (cooldown >= 0.99f && timeSinceLastAttack >= minDelay) {
                    currentState = State.ATTACKING
                }
            }
            State.ATTACKING -> {
                if (target == null || !isValidTarget(target)) {
                    currentState = State.IDLE
                    return
                }
                
                attackEntity(target)
                lastAttackTime.set(System.currentTimeMillis())
                currentState = State.ROTATING
            }
        }
    }
    
    private fun scanForTargets() {
        val mc = MinecraftClient.getInstance()
        val player = mc.player ?: return
        val world = mc.world ?: return
        
        val playerPos = Vec3d(player.x, player.y, player.z)
        val box = Box.of(playerPos, range.toDouble() * 2, range.toDouble() * 2, range.toDouble() * 2)
        val entities = world.getOtherEntities(player, box) { entity ->
            entity is LivingEntity && 
            entity.isAlive && 
            player.distanceTo(entity) <= range &&
            (!playersOnly || entity is PlayerEntity)
        }
        
        val closest = entities
            .filterIsInstance<LivingEntity>()
            .filter { isValidTarget(it) }
            .minByOrNull { player.distanceTo(it) }
        
        currentTarget.set(closest)
    }
    
    private fun isValidTarget(target: LivingEntity): Boolean {
        val mc = MinecraftClient.getInstance()
        val player = mc.player ?: return false
        
        if (!target.isAlive || target.health <= 0f) return false
        if (player.distanceTo(target) > range) return false
        if (!throughWalls && !hasLineOfSight(target)) return false
        
        return true
    }
    
    private fun hasLineOfSight(target: Entity): Boolean {
        val mc = MinecraftClient.getInstance()
        val player = mc.player ?: return false
        val world = mc.world ?: return false
        
        val eyePos = Vec3d(player.x, player.eyeY, player.z)
        val targetPos = Vec3d(target.x, target.y + target.standingEyeHeight.toDouble() * 0.5, target.z)
        val direction = targetPos.subtract(eyePos).normalize()
        val distance = eyePos.distanceTo(targetPos)
        
        val context = RaycastContext(
            eyePos,
            eyePos.add(direction.multiply(distance, distance, distance)),
            RaycastContext.ShapeType.OUTLINE,
            RaycastContext.FluidHandling.NONE,
            player
        )
        
        val result = world.raycast(context)
        return result == null || result.type == net.minecraft.util.hit.HitResult.Type.MISS ||
               result.pos.distanceTo(eyePos) >= distance - 0.1
    }
    
    private fun calculateIdealRotations(target: Entity): Pair<Float, Float> {
        val mc = MinecraftClient.getInstance()
        val player = mc.player ?: return Pair(0f, 0f)
        
        val eyePos = Vec3d(player.x, player.eyeY, player.z)
        val targetPos = Vec3d(target.x, target.y + target.standingEyeHeight.toDouble() * 0.5, target.z)
        val diff = targetPos.subtract(eyePos)
        
        val distanceXZ = sqrt(diff.x * diff.x + diff.z * diff.z)
        val yaw = Math.toDegrees(atan2(diff.z, diff.x)).toFloat() - 90f
        val pitch = -Math.toDegrees(atan2(diff.y, distanceXZ)).toFloat()
        
        return Pair(normalizeYaw(yaw), pitch.coerceIn(-90f, 90f))
    }
    
    private fun updateRotations(target: Entity) {
        val (idealYaw, idealPitch) = calculateIdealRotations(target)
        serverYaw = normalizeYaw(serverYaw)
        
        var deltaYaw = normalizeYaw(idealYaw - serverYaw)
        val noise = Random.nextFloat() * 0.2f - 0.1f
        deltaYaw += noise
        
        val yawStep = (rotationSpeed * easeInOutCubic(abs(deltaYaw) / 180f)).coerceIn(1f, rotationSpeed)
        
        if (abs(deltaYaw) > yawStep) {
            serverYaw += if (deltaYaw > 0) yawStep else -yawStep
        } else {
            serverYaw = idealYaw
        }
        
        var deltaPitch = idealPitch - serverPitch
        deltaPitch += noise * 0.5f
        
        val pitchStep = (rotationSpeed * easeInOutCubic(abs(deltaPitch) / 90f)).coerceIn(1f, rotationSpeed)
        
        if (abs(deltaPitch) > pitchStep) {
            serverPitch += if (deltaPitch > 0) pitchStep else -pitchStep
        } else {
            serverPitch = idealPitch
        }
        
        serverPitch = serverPitch.coerceIn(-90f, 90f)
        
        if (Random.nextFloat() < 0.05f) {
            isRotating = false
            return
        }
        
        isRotating = true
    }
    
    private fun sendRotations() {
        val mc = MinecraftClient.getInstance()
        val player = mc.player ?: return
        val networkHandler = mc.networkHandler ?: return
        
        val packet = PlayerMoveC2SPacket.LookAndOnGround(
            serverYaw,
            serverPitch,
            player.isOnGround,
            player.horizontalCollision
        )
        
        networkHandler.sendPacket(packet)
    }
    
    private fun attackEntity(target: Entity) {
        val mc = MinecraftClient.getInstance()
        val player = mc.player ?: return
        val networkHandler = mc.networkHandler ?: return
        
        val attackPacket = PlayerInteractEntityC2SPacket.attack(target, player.isSneaking)
        networkHandler.sendPacket(attackPacket)
        
        if (Random.nextFloat() > 0.1f) {
            val swingPacket = HandSwingC2SPacket(Hand.MAIN_HAND)
            networkHandler.sendPacket(swingPacket)
        }
        
        player.resetLastAttackedTicks()
    }
    
    private fun normalizeYaw(yaw: Float): Float {
        var normalized = yaw % 360f
        if (normalized > 180f) normalized -= 360f
        if (normalized < -180f) normalized += 360f
        return normalized
    }
    
    private fun easeInOutCubic(t: Float): Float {
        return if (t < 0.5f) {
            4f * t * t * t
        } else {
            1f - (-2f * t + 2f).pow(3) / 2f
        }
    }
    
    fun setRange(newRange: Float) {
        range = newRange.coerceIn(3.0f, 6.0f)
    }
    
    fun setCps(newCps: Float) {
        cps = newCps.coerceIn(8f, 20f)
    }
    
    fun setRotationSpeed(newSpeed: Float) {
        rotationSpeed = newSpeed.coerceIn(5f, 30f)
    }
    
    fun setPlayersOnly(value: Boolean) {
        playersOnly = value
    }
    
    fun setThroughWalls(value: Boolean) {
        throughWalls = value
    }
}